### PR TITLE
Feature/306 browser groups

### PR DESCRIPTION
**Pull request by zuevval**
The implementation of #306

screenshot  screenshot

### CODE REVIEW

**Comment by winter-yuki** [comment]
Pause your development here please, it looks like a suboptimal way now

File: app/src/main/java/com/github/braillesystems/learnbraille/data/repository/MaterialsRepository.kt [hunk 1]
```diff
@@ -44,4 +49,7 @@ open class MaterialsRepositoryImpl(
         } else {
             deckDao.allDecks().map { DeckWithAvailability(it, true) }
         }
+
+    override suspend fun allDecksWithUniqueMaterials(): List<Deck> = DeckTags.Unique.values()
+        .mapNotNull { deckDao.deckByTag(it.tag) }
```
Comment (winter-yuki) [inline app/src/main/java/com/github/braillesystems/learnbraille/data/repository/MaterialsRepository.kt:54]: A database query is executed for each Unique tag. It is significantly more efficient to grab all decks from the database and filter them.

File: app/src/main/java/com/github/braillesystems/learnbraille/ui/screens/browser/BrowserDeckFragment.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ) = DataBindingUtil.inflate<FragmentBrowserBinding>(
+        inflater,
+        R.layout.fragment_browser,
+        container,
+        false
+    ).ini().also { binding ->
+        lifecycleScope.launch {
+            val deckId = browserRepository.currentDeckId
```
Comment (winter-yuki) [inline app/src/main/java/com/github/braillesystems/learnbraille/ui/screens/browser/BrowserDeckFragment.kt:41]: It is much better to pass the current deck via fragment args than via the persistent global state.

Comment (winter-yuki) [reply app/src/main/java/com/github/braillesystems/learnbraille/ui/screens/browser/BrowserDeckFragment.kt:41]: I'll fix this

**Review by winter-yuki — CHANGES_REQUESTED** [review]
I need to make some changes myself

**Comment by zuevval** [comment]
> I need to make some changes myself

OK, thanks for the review. In this case, I’d rather wait for your changes and then fix the remaining issues

**Comment by winter-yuki** [comment]
I do have not enough time to carefully refactor now. Can you please fix some aforementioned issues and merge PR?

**Comment by zuevval** [comment]
@winter-yuki OK, I will take a look (most likely on Sunday 12th)

**Comment by zuevval** [comment]
OK, I failed to complete this by today, but I'll work on it tomorrow

File: app/src/main/java/com/github/braillesystems/learnbraille/res/Data.kt [hunk 3]
```diff
@@ -118,35 +119,42 @@ object StepAnnotation {
...
+    enum class Unique(val tag: String) {
+        RuLetters("ru_letters"),
+        LatinLetters("latin_letters"),
+        GreekLetters("greek_letters"),
+        Digits("digits"),
+        Markers("markers"),
+        Special("special"),
+        Math("math"),
+        MusicalNotes("notes"),
+        OtherMusic("otherMusic")
```
[the remark this answers is not in the document]
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/res/Data.kt]: fixed: "otherMusic" -> "other_music"

File: app/src/main/java/com/github/braillesystems/learnbraille/res/Data.kt [hunk 4]
```diff
@@ -118,35 +119,42 @@ object StepAnnotation {
...
-    const val ruLetters = "ru_letters"
-    const val latinLetters = "latin_letters"
-    const val greekLetters = "greek_letters"
-    const val digits = "digits"
-    const val markers = "markers"
-    const val special = "special"
-    const val math = "math"
-    const val musical_notes = "notes"
-    const val other_music = "other_music"
+    object Grouping {
```
[the remark this answers is not in the document]
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/res/Data.kt]: Changed to enum

File: app/src/main/java/com/github/braillesystems/learnbraille/res/Data.kt [hunk 5]
```diff
@@ -118,35 +119,42 @@ object StepAnnotation {
...
+        Special("special"),
+        Math("math"),
+        MusicalNotes("notes"),
+        OtherMusic("otherMusic")
+    }
 }
 
 val Context.deckTagToName: Map<String, String> by lazyWithContext {
-    DeckTags.run {
+    val generalizing = DeckTags.Grouping.run {
```
[the remark this answers is not in the document]
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/res/Data.kt]: changed to a single map

File: app/src/main/java/com/github/braillesystems/learnbraille/data/repository/BrowserRepository.kt [hunk 6]
```diff
@@ -6,7 +6,7 @@ import com.github.braillesystems.learnbraille.data.entities.DBid
 import com.github.braillesystems.learnbraille.utils.preferences
 
 interface BrowserRepository : MaterialsRepository {
-    val currentDeckId: DBid
+    var currentDeckId: DBid
```
[the remark this answers is not in the document]
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/data/repository/BrowserRepository.kt]: As I can see, you've fixed it some time after this comment

File: app/src/main/java/com/github/braillesystems/learnbraille/data/repository/MaterialsRepository.kt [hunk 7]
```diff
@@ -1,19 +1,22 @@
...
 
 interface MaterialsRepository {
+    suspend fun deck(id: DBid): Deck?
     suspend fun randomMaterialFromDeck(id: DBid): Material?
     suspend fun randomKnownMaterialFromDeck(id: DBid): Material?
     suspend fun allMaterialsFromDeck(id: DBid): List<Material>
     suspend fun allDecks(): List<Deck>
     suspend fun availableDecks(): List<Deck>
     suspend fun allDecksWithAvailability(): List<DeckWithAvailability>
+    suspend fun allDecksWithUniqueMaterials(): List<Deck>
```
[the remark this answers is not in the document]
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/data/repository/MaterialsRepository.kt]: done

File: app/src/main/java/com/github/braillesystems/learnbraille/data/repository/MaterialsRepository.kt (hunk 1 above)
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/data/repository/MaterialsRepository.kt]: fixed this

File: app/src/main/java/com/github/braillesystems/learnbraille/ui/screens/browser/BrowserDeckFragment.kt (hunk 2 above)
Comment (zuevval) [reply app/src/main/java/com/github/braillesystems/learnbraille/ui/screens/browser/BrowserDeckFragment.kt:41]: Maybe in the next life

**Pull request merged**